### PR TITLE
Cw master

### PR DESCRIPTION
**Pull request by CarloWood**
In the past days I reviewed cmake configuration (sitl in particular) in order to understand how to debug a simulation (needed because it core dumps for me at start up).

These are the commits that I made so far, with in mind public use by others; it involves some cleanup, minor bug fixes and in the end a commit that allows a developer to specify which targets (executables and/or modules) he or she wants to suppress compiler optimization for.

### CODE REVIEW

File: cmake/common/px4_base.cmake [hunk 1]
```diff
@@ -743,20 +753,21 @@ function(px4_add_common_flags)
...
 	set(added_cxx_flags
 		${cxx_compile_flags}
 		${warnings}
 		${cxx_warnings}
-		${max_optimization}
-		${optimization_flags}
 		${visibility_flags}
 		)
 
+	set(added_optimization_flags
```
Comment (bkueng) [inline cmake/common/px4_base.cmake]: I don't see this being used anywhere

**Comment by bkueng** [comment]
Great! Generally we merge cmake files as much as possible, so your changes are much appreciated.

- Merging the targets in the Makefile is a good idea. One drawback I see is that it's not easy to see which targets exist. I think it's time to add a `make help` (I can take care of that). For the replay cmake file: this is for our old logger, which is in the process of being replaced, together with a new replay method.
- CI fails because of an overflown flash, which indicates that some flags got changed/dropped. Can you look into that?
- A minor thing: we use a rebase workflow, so instead of merging master into a feature branch, we rebase the branch onto master.
- `find . -wholename '*/CMakeFiles/*.dir/flags.make' | xargs dirname | xargs basename -a | sort -u | sed -e 's/.dir$//'` does not work if `ninja` is used
- Can you add documentation for the optimization flags to http://dev.px4.io/simulation-debugging.html?
- just an fyi: most of the modules add a `-Os` flag, which is unnecessary, and I plan to remove these

**Comment by davids5** [comment]
@CarloWood, @bkueng 

Merging targets: We will need the completely enumerated targets list a list form - without abbreviations

With Target specific optimization we should also consider configurable stack sizing based on a target.

@bkueng Per module optimization is nice to have when debugging - so let's not lose that ability

**Comment by bkueng** [comment]
@davids5 Why do we need a complete list?
I wasn't going to remve the per-module optimization ability, just the currently repeated flag that is set already in the base config.

**Comment by davids5** [comment]
@bkueng Please see this https://github.com/PX4/Firmware/issues/5298

**Comment by davids5** [comment]
Wrong button!

File: cmake/common/px4_base.cmake (hunk 1 above)
Comment (CarloWood) [inline cmake/common/px4_base.cmake]: Hi bkueng. The added_ variables are used by reconstructing them (as ${added_${lower_var}}), a bit lower (line 1106). I overlooked this at first too! ;)

**Comment by CarloWood** [comment]
@bkueng wrote:
* Merging the targets in the Makefile is a good idea. One drawback I see is that it's not easy to see which targets exist. I think it's time to add a make help (I can take care of that). For the replay cmake file: this is for our old logger, which is in the process of being replaced, together with a new replay method.

I think a list of all viewers, models and debuggers in the main Makefile would be good,
which then should be propagated down to other (cmake) files that need them, in order to avoid duplicated maintenance.

* CI fails because of an overflown flash, which indicates that some flags got changed/dropped. Can you look into that?

My last commit fixes this - I forgot that now calls to add_executable and add_library that were not replaced with px4_add_executable / px4_add_library now lost their optimization flags %-)

* A minor thing: we use a rebase workflow, so instead of merging master into a feature branch, we rebase the branch onto master.

What does this mean for me? Do I have to resubmit this pull request?
I'm used to use git as it was intended :p not as a glorified svn </religious_war>

* find . -wholename '*/CMakeFiles/*.dir/flags.make' | xargs dirname | xargs basename -a | sort -u | sed -e 's/.dir$//' does not work if ninja is used

Hence a make target that prints them would be nice...

* Can you add documentation for the optimization flags to http://dev.px4.io/simulation-debugging.html?
just an fyi: most of the modules add a -Os flag, which is unnecessary, and I plan to remove these

Yes, I noticed that - they are completely unnecessary - but I was already changing so much that I didn't want to rock the boat in too many ways :p. I'm glad you already realize it too.
I will have a look at that url and come up with some documentation.

**Comment by CarloWood** [comment]
@davids5 wrote

* With Target specific optimization we should also consider configurable stack sizing based on a target.

Yes, I also realized this. Especially since the optimization suppression is only really useful during simulation, in which case we're not really limited in available memory. Overflows occurring during simulation while suppressing optimization should be a non-issue and still just work.

**Comment by davids5** [comment]
Hi @CarloWood 

In addition to all that you have here, there are  a few areas that can use improvement as well. Things that should only be driven by the makefile or only have one instance of data. Take for instance, the Submodule list lives in a script in tools - there are more things like this too.

**Comment by CarloWood** [comment]
I wrote some docs, what about this:
https://github.com/CarloWood/Devguide/blob/d8c8bd418e6fcef71432a5511a5416e264f5c1f9/simulation-debugging.md

**Comment by CarloWood** [comment]
I am going to rename my branch from cw_master to PR5432 and then push again... lets see if github likes that for a branch that already has a pull-request ;).

**Comment by CarloWood** [comment]
Renaming caused this pull request not to be synced anymore.. so I renamed it back.

**Comment by CarloWood** [comment]
I'm closing this pull request because I did a rebase onto master for a branch with a better name.

**Comment by CarloWood** [comment]
New pull request is now here: https://github.com/PX4/Firmware/pull/5443

**Pull request closed without merging**